File: app/src/main/java/io/plan8/business/adapter/BindingRecyclerViewAdapter.java
package io.plan8.business.adapter;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by SSozi on 2017. 9. 15..
 */

public abstract class BindingRecyclerViewAdapter<T> extends RecyclerView.Adapter<BindingRecyclerViewAdapter.BindingViewHolder> {
    protected List<T> data = new ArrayList<>();

    abstract protected int selectViewLayoutType(T data);

    abstract protected void bindVariables(ViewDataBinding binding, T data);

    public List<T> getData() {
        return data;
    }

    public void setData(List<T> data) {
        if (data != null) this.data = data;
        else this.data = new ArrayList<>();
        notifyDataSetChanged();
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public BindingRecyclerViewAdapter.BindingViewHolder onCreateViewHolder(ViewGroup parent, int layoutId) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(layoutId, parent, false);
        return new BindingViewHolder(view);
    }

    @Override
    public void onBindViewHolder(BindingRecyclerViewAdapter.BindingViewHolder viewHolder, int position) {
        bindVariables(viewHolder.getBinding(), data.get(position));
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @Override
    public int getItemViewType(int position) {
        return selectViewLayoutType(data.get(position));
    }

    public class BindingViewHolder extends RecyclerView.ViewHolder {
        private ViewDataBinding binding;

        public BindingViewHolder(View view) {
            super(view);
            this.binding = DataBindingUtil.bind(view);
        }

        public ViewDataBinding getBinding() {
            return binding;
        }
    }
}

File: app/src/main/java/io/plan8/business/adapter/BindingRecyclerViewAdapter.kt
package io.plan8.business.adapter

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.util.*

/**
 * Created by SSozi on 2017. 9. 15..
 */

abstract class BindingRecyclerViewAdapter<T> : RecyclerView.Adapter<BindingRecyclerViewAdapter.BindingViewHolder>() {
    var data: List<T>? = null
        set (data) {
            if (data != null)
                field = data
            else
                field = ArrayList()
            notifyDataSetChanged()
        }

    protected abstract fun selectViewLayoutType(data: T): Int

    protected abstract fun bindVariables(binding: ViewDataBinding, data: T)

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        super.onAttachedToRecyclerView(recyclerView)
    }

    override fun onCreateViewHolder(parent: ViewGroup, layoutId: Int): BindingRecyclerViewAdapter.BindingViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(layoutId, parent, false)
        return BindingViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: BindingRecyclerViewAdapter.BindingViewHolder, position: Int) {
        bindVariables(viewHolder.binding, data!![position])
    }

    override fun getItemCount(): Int {
        return data!!.size
    }

    override fun getItemViewType(position: Int): Int {
        return selectViewLayoutType(data!![position])
    }

    class BindingViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val binding: ViewDataBinding

        init {
            this.binding = DataBindingUtil.bind(view)
        }
    }
}